src/render/tests/test_freeze.py

```diff
@@ -0,0 +1,1170 @@
+    for ref, frozen in zip(images_ref, images_frozen):
+        assert dr.allclose(ref, frozen)
+
+
+@pytest.mark.parametrize(
+    "integrator",
+    [
+        "direct",
+        "path",
+        "prb",
+        "prb_basic",
+        "direct_projective",
+        "prb_projective",
+        "moment",
+        "ptracer",
+        "depth",
+    ],
+)
+def test06_integrators(variants_vec_rgb, integrator):
+    # dr.set_log_level(dr.LogLevel.Trace)
+    # dr.set_flag(dr.JitFlag.ReuseIndices, False)
+    # dr.set_flag(dr.JitFlag.Debug, True)
+
+    w = 16
+    h = 16
+
+    n = 5
+
+    def func(scene: mi.Scene) -> mi.TensorXf:
+        with dr.profile_range("render"):
+            result = mi.render(scene, spp=1)
+        return result
+
+    frozen = dr.freeze(func)
+
+    def load_scene():
+        scene = mi.cornell_box()
+        scene["sensor"]["film"]["width"] = w
+        scene["sensor"]["film"]["height"] = h
+
+        if integrator == "path":
+            scene["integrator"] = {
+                "type": "path",
+                "max_depth": 4,
+            }
+        elif integrator == "direct":
+            scene["integrator"] = {
+                "type": "direct",
+            }
+        elif integrator == "prb":
+            scene["integrator"] = {
+                "type": "prb",
+            }
+        elif integrator == "prb_basic":
+            scene["integrator"] = {
+                "type": "prb_basic",
+            }
+        elif integrator == "direct_projective":
+            scene["integrator"] = {
+                "type": "direct_projective",
+            }
+        elif integrator == "prb_projective":
+            scene["integrator"] = {
+                "type": "prb_projective",
+            }
+        elif integrator == "moment":
+            scene["integrator"] = {
+                "type": "moment",
+                "nested": {
+                    "type": "path",
+                },
+            }
+        elif integrator == "ptracer":
+            scene["integrator"] = {
+                "type": "ptracer",
+                "max_depth": 8,
+            }
+        elif integrator == "depth":
+            scene["integrator"] = {"type": "depth"}
+
+        scene = mi.load_dict(scene, parallel=False)
+        return scene
+
+    def run(n: int, func: Callable[[mi.Scene], mi.TensorXf]) -> List[mi.TensorXf]:
+        scene = load_scene()
+
+        images = []
+        for i in range(n):
+            img = func(scene)
+            dr.eval(img)
+
+            images.append(img)
+
+        return images
+
+    # scene = load_scene()
+    images_ref = run(n, func)
+    images_frozen = run(n, frozen)
+
+    assert frozen.n_recordings < n
+    for ref, frozen in zip(images_ref, images_frozen):
+        assert dr.allclose(ref, frozen)
+
+
+@pytest.mark.parametrize(
+    "shape",
+    [
+        "mesh",
+        "disk",
+        "cylinder",
+        "bsplinecurve",
+        "linearcurve",
+        "sdfgrid",
+        # "instance",
```

Comment:
Are `instance` shapes not supported?
If we don't manage to supported them, we should find a way to throw an exception if the user attempts to use them.